### PR TITLE
Issue/268

### PR DESCRIPTION
**Pull request by minborg**
Expose an internal compare method and add doc and tests

### CODE REVIEW

**Comment by minborg** [comment]
It appears this branch was based on another branch rather than `develop`. Sorry.

**Comment by RobAustin** [comment]
image

File: src/test/java/net/openhft/chronicle/bytes/BytesFactoryUtil.java [hunk 1]
```diff
@@ -211,4 +211,65 @@ public void accept(T t) {
     }
 
 
+    static final class BytesInitialInfo {
+        private final long readPosition;
+        private final long writePosition;
+        private final long writeLimit;
+        private final long capacity;
+        private final long realCapacity;
+        //
```
Comment (RobAustin) [inline src/test/java/net/openhft/chronicle/bytes/BytesFactoryUtil.java]: why the //

File: src/test/java/net/openhft/chronicle/bytes/BytesReleaseInvariantNonPerformantMethodsTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        final OutputStream os = new OutputStream() {
+            @Override
+            public void write(int b) throws IOException {
+                throw new UnsupportedEncodingException();
+            }
+        };
+        final BytesStore<?, ?> bs = BytesStore.from(SILLY_NAME);
+        final Bytes<?> bytes = Bytes.from(SILLY_NAME);
+        return Stream.of(
+                NamedConsumer.of(Bytes::toHexString, "toHexString()"),
```
Comment (RobAustin) [inline src/test/java/net/openhft/chronicle/bytes/BytesReleaseInvariantNonPerformantMethodsTest.java]: consider using static import for the NamedConsumer

File: src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java [hunk 3]
```diff
@@ -265,7 +266,7 @@ public void testCopyTo() {
         final Bytes<ByteBuffer> dst = Bytes.elasticByteBuffer();
         try {
             dst.writePosition(src.copyTo(dst));
-            assertEquals(src, dst);
+            assertTrue(BytesInternal.contentEqual(src, dst));
```
Comment (RobAustin) [inline src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java]: suggestion -use AssertEquals(,) but first convert both to a string as this gives you a better diff in the IDE if it fails

File: pom.xml [hunk 4]
```diff
@@ -26,7 +26,7 @@
     </parent>
 
     <artifactId>chronicle-bytes</artifactId>
-    <version>2.22ea11-SNAPSHOT</version>
+    <version>2.22ea12-SNAPSHOT</version>
```
Comment (JerryShea) [inline pom.xml]: why is this changed? Do we need to rebase?

**Review by nicktindall — COMMENTED** [review]
Would it be better if we had a pattern like 

```
try (Reference ref = bytes.acquireReferenceOrThrowIfReleased()) {
   ... do stuff
}
```
instead of `throwExceptionIfReleased`, that way we'd be protected against the thing being released while we're in the block as well? I think we discussed something similar in the dev meeting?

It would also have the behaviour that if you were passed a released bytes it would throw.

File: src/test/java/net/openhft/chronicle/bytes/BytesReleaseInvariantNonPerformantMethodsTest.java (hunk 2 above)
Comment (JerryShea) [reply src/test/java/net/openhft/chronicle/bytes/BytesReleaseInvariantNonPerformantMethodsTest.java]: @RobAustin strongly disagree - lots of things have an "of" method - `NamedConsumer.of` reads nicely and is clear

File: src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java (hunk 3 above)
Comment (minborg) [reply src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java]: Because we changed how `Bytes::equals` works (it might consider the contents or it might use `==`) then this is the way to go. We might create a custom method (for testing) that asserts content equality and that shows the difference.

Comment (minborg) [reply src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java]: Ahh. I did not read your comment well enough. Good suggestion.

File: src/test/java/net/openhft/chronicle/bytes/BytesFactoryUtil.java (hunk 1 above)
Comment (minborg) [reply src/test/java/net/openhft/chronicle/bytes/BytesFactoryUtil.java]: No idea... I will remove

**Comment by minborg** [comment]
I have promoted the comment of @nicktindall to an issue: https://github.com/OpenHFT/Chronicle-Bytes/issues/273

File: src/main/java/net/openhft/chronicle/bytes/internal/BytesInternal.java [hunk 5]
```diff
@@ -139,10 +139,16 @@ static boolean contentEqualInt(BytesStore a, BytesStore b)
         }
     }
 
-    public static boolean contentEqual(@Nullable BytesStore a, @Nullable BytesStore b)
-            throws IllegalStateException {
+    public static boolean contentEqual(@Nullable final BytesStore a,
```
Comment (RobAustin) [inline src/main/java/net/openhft/chronicle/bytes/internal/BytesInternal.java:142]: please add java doc

File: src/main/java/net/openhft/chronicle/bytes/BytesUtil.java [hunk 6]
```diff
@@ -151,6 +152,29 @@ public static void writeFile(String file, Bytes<byte[]> bytes)
...
+     * @param a a ByteStore
+     * @param b a ByteStore to be compared with {@code a} for equality
+     * @return if the arguments are equal to each other
+     * @throws ClosedIllegalStateException if either of the provided parameters are released
+     *
+     * @see Object#equals(Object)
+     */
+    public static boolean contentEqual(@Nullable final BytesStore a,
+                                       @Nullable final BytesStore b) {
+        return BytesInternal.contentEqual(a, b);
```
Comment (RobAustin) [inline src/main/java/net/openhft/chronicle/bytes/BytesUtil.java:175]: does this throw IllegalStateException ?  javadoc refers to ClosedIllegalStateException

**Review by RobAustin — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/bytes/BytesUtil.java (hunk 6 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/bytes/BytesUtil.java:175]: It throws `ClosedIllegalStateException` as specified in the JavaDocs. 

```
    public static void throwExceptionIfReleased(final ReferenceCounted referenceCounted) {
        if (referenceCounted.refCount() <= 0) {
            throw new ClosedIllegalStateException("Released");
        }
    }
```

File: src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java (hunk 3 above)
Comment (RobAustin) [reply src/test/java/net/openhft/chronicle/bytes/NativeBytesStoreTest.java]: > Ahh. I did not read your comment well enough. Good suggestion.

are you gonna change it again ?
